orbit/pkg/service/service_windows.go

```diff
@@ -0,0 +1,95 @@
+//go:build windows
+// +build windows
+
+package service
+
+import (
+	"os"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/orbit/pkg/constant"
+	"github.com/fleetdm/fleet/v4/orbit/pkg/platform"
+
+	"github.com/rs/zerolog/log"
+	"golang.org/x/sys/windows/svc"
+)
+
+type windowsService struct {
+	rootDir string
+}
+
+func (m *windowsService) bestEffortShutdown(serviceRootDir string) {
+	err := platform.KillProcessByName(constant.DesktopAppExecName)
+	if err != nil {
+		log.Info().Err(err).Msg("The desktop app couldn't be killed")
+	}
+
+	if serviceRootDir != "" {
+		err = platform.KillFromPIDFile(serviceRootDir, constant.OsqueryPidfile, constant.OsquerydName)
+		if err != nil {
+			log.Info().Err(err).Msg("The osquery daemon process couldn't be killed")
+		}
+	}
+}
+
+func (m *windowsService) Execute(args []string, requests <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	// Accepted service operations
+	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
+
+	// Expected service status update during initialization
+	changes <- svc.Status{State: svc.StartPending}
+	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
+
+	// The listening loop below will keep listening for new SCM requests
+	for req := range requests {
+		switch req.Cmd {
+
+		case svc.Interrogate:
+			changes <- req.CurrentStatus
+
+		case svc.Stop, svc.Shutdown:
+
+			// Service shutdown was requested
+			// Updating the service state to indicate stop op started
+			changes <- svc.Status{State: svc.StopPending}
+
+			// Best effort tear down
+			m.bestEffortShutdown(m.rootDir)
+
+			// Updating the service state to indicate that stop finished
+			changes <- svc.Status{State: svc.Stopped}
+
+			// Dummy delay to allow the SCM to pick up the changes
+			time.Sleep(500 * time.Millisecond)
+
+			// Drastic teardown
+			os.Exit(0)
+
+		default:
+			return false, 1052 // 1052: ERROR_INVALID_SERVICE_CONTROL
+		}
+	}
+
+	return false, 0
+}
+
+// This function implements the dispatcher and notification logic to
+// interact with the Windows Service Control Manager (SCM)
+func SetupServiceManagement(serviceName string, serviceRootDir string) {
+	if (serviceName != "") && (serviceRootDir != "") {
+		// Ensuring that we are only calling the SCM if running as a service
+		isWindowsService, err := svc.IsWindowsService()
+		if err == nil {
+			if isWindowsService {
+				srvData := windowsService{rootDir: serviceRootDir}
+				// Registering our service into the SCM
+				err := svc.Run(serviceName, &srvData)
+				if err != nil {
+					log.Info().Err(err).Msg("SCM registration failed")
+				}
+			}
+		} else {
+			log.Info().Err(err).Msg("couldn't determine if running as a service")
+		}
```

Comment:
We usually prefer the "early return" pattern for error handling in Go:

```
if err != nil {
    // log
    return // (or return error depending on signature of function)
}
// continue execution 
```

Reply:
Hey @zwass, so I found this behavior was not deterministic. It seems that only happens if the service is stopped in the middle of an update (which causes the update interrupt() function to be called). So I modified this, and I'm now saving the interrupt functions into a slice and passing them to the service shutdown helper so interrupts can be properly called to gracefully shutdown the process. Only on interrupt() call is required to have group runner call the rest of the interrupt functions